feat: parse lists

ematipico opened the pull request:
<!--
	Thanks for submitting a pull request!

	We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request.

	Once created, your PR will be automatically labeled according to changed files.

	Learn more about contributing: https://github.com/rome/tools/blob/main/CONTRIBUTING.md
-->

## Summary

This PR closes #1836 

I created two traits - for now - that implement two different kind of lists:
- `ParseNodeList`: normal lists, nothing fancy
- `ParseSeparatedList`: lists separated by a recurring separating element, which will be always the same, e.g.: `["foo", 1, true]`, where `,` is the recurring separating element


`ParseNodeList` and `ParseSeparatedList` have the same API and they are not shared.
Myself and @MichaReiser agreed on the fact that it's better to make the developer experience easier instead of making maintainability a priority.


The PR moves one of the lists to this new methodology and there should not be regressions. It also shows how would they work. 

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->

## Test Plan

Current CI

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output. -->

MichaReiser commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,183 @@
+pub struct ParseList<Condition, CreateElement> {
+	/// This is a condition used inside the loop. When it's false, it exists the loop
+	condition: Option<Condition>,
+	/// A closure that is charge of creating the element inside the loop
+	create_element: Option<CreateElement>,
```
Nit: `create` sounds the element is created out of the blue (just a new element) but what it's doing is to `parse` an element. That's why I would rename it to `parse_element`.

MichaReiser commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,183 @@
+		}
+	}
+
+	#[must_use = "You need to provide the condition to exit from the loop"]
```
[`must_use`](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute) specifies that the returned value must be used, but not that a certain method must be called. 

So, the `must_use` protects` here from the case where the returned builder is not used, but that's already guaranteed by the compiler, because the method "consumes" the builder.

MichaReiser commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,183 @@
+		let mut progress = ParserProgress::default();
+		while (condition)(p) {
+			progress.assert_progressing(p);
+			(create_element)(p).or_missing(p);
```
A list shouldn't contain "missing" elements, except if it is a separated list. 

It's also unsafe to call `or_missing` without any recovery because the parser didn't progress when parsing the element but won't progress in the next loop as well because it's still at the same position.

That's why I don't think we should offer a `parse_list` function without any recovery and a parse function that returns `ParsedSyntax`. No recovery and a `parse` method that always returns an item or recovers.

MichaReiser reviewed (commented):
I saw that you requested feedback which is why I did a review. I hope that's what you intended. 

I'm not sure what the rustiest way of implementing `parse_list` is but using traits worked well when implementing the `ParsePattern` helpers. 

The reason why traits could work well is that we may be able to implement a generic `ParseList` trait that implements the "loop" but delegates most to methods that different trait implementations can override if they want. 

The main downside is that it now requires a struct implementing that trait, which is slightly more code but we could get away with helper methods like `parse_array` that use good defaults and have `setter` methods to customise the behaviour. 

```rust
trait ParseList<T> {
 fn parse_list(&mut self, p: &mut Parser) -> ParsedSyntax {
  // TODO should this also eat the open / close brackets? 
  let list = self.start_list(p); // Or just call p.start()? 
  let mut progress = ParserProgress:default();
  
 while (!p.at(EOF) && !self.is_at_end(p)) {
     progress.assert_progressing(p);
     let item = self.parse_item(p);
     
     if item.is_absent() && self.recover(p).is_err()  {
       break;
     }
  }
  
  self.finish_list(p, list)
 }
 
 fn recover(&mut self, p: &mut Parser) -> RecoveryResult;
 fn start_list(&mut self, p: &mut Parser) -> Marker;
 fn finish_list(&mut self, p: &mut Parser) -> ParsedSyntax<CompletedMarker>;
 fn is_at_end(&mut self, p: &mut Parser) -> bool;
}

```

I believe that this could be flexible enough to even implement parsing for a separated list because the list is allowed to track some state internally. 

`T` so that we can implement a `ParseList<ParsedSyntax>` and `ParseList<ConditionalSyntax>`?

ematipico commented:
> I saw that you requested feedback which is why I did a review. I hope that's what you intended.
> 

Yes, that's exactly what I wanted, thank you! :) 

> I'm not sure what the rustiest way of implementing `parse_list` is but using traits worked well when implementing the `ParsePattern` helpers.
> 
> The reason why traits could work well is that we may be able to implement a generic `ParseList` trait that implements the "loop" but delegates most to methods that different trait implementations can override if they want.
> 

That was my other solution 😅

> The main downside is that it now requires a struct implementing that trait, which is slightly more code but we could get away with helper methods like `parse_array` that use good defaults and have `setter` methods to customise the behaviour.
> 
> ```rust
> trait ParseList<T> {
>  fn parse_list(&mut self, p: &mut Parser) -> ParsedSyntax {
>   // TODO should this also eat the open / close brackets? 
>   let list = self.start_list(p); // Or just call p.start()? 
>   let mut progress = ParserProgress:default();
>   
>  while (!p.at(EOF) && !self.is_at_end(p)) {
>      progress.assert_progressing(p);
>      let item = self.parse_item(p);
>      
>      if item.is_absent() && self.recover(p).is_err()  {
>        break;
>      }
>   }
>   
>   self.finish_list(p, list)
>  }
>  
>  fn recover(&mut self, p: &mut Parser) -> RecoveryResult;
>  fn start_list(&mut self, p: &mut Parser) -> Marker;
>  fn finish_list(&mut self, p: &mut Parser) -> ParsedSyntax<CompletedMarker>;
>  fn is_at_end(&mut self, p: &mut Parser) -> bool;
> }
> ```
> 
> I believe that this could be flexible enough to even implement parsing for a separated list because the list is allowed to track some state internally.
> 
> `T` so that we can implement a `ParseList<ParsedSyntax>` and `ParseList<ConditionalSyntax>`?

xunilrj commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,183 @@
+		self
+	}
+
+	#[must_use = "You need provide the function that creates the single element list"]
```
Same thing here about @MichaReiser comment.

xunilrj commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,183 @@
+	/// When using it, if [create_element] returns a [ParsedSyntax::Absent], it will
+	/// mark it as a missing element
+	pub fn create_list(self, p: &mut Parser) -> ParsedSyntax<CompletedMarker> {
+		let condition = self
```
If you use the builder pattern I suggested above, you will probably not need to check this, because this field will not be ```Option<>```.

xunilrj commented on crates/rslint_parser/src/syntax/stmt.rs:
```diff
@@ -1105,14 +1105,13 @@ fn parse_switch_clause(
-			cons_list.complete(p, LIST);
+			ParseList::new()
+				.set_condition(|p| !p.at_ts(token_set![T![default], T![case], T!['}'], EOF]))
+				.set_create_element(|p| Present(parse_statement(p, None).unwrap()))
```
We can do

```rust
.set_create_element(|p| parse_statement(p, None).into())
```

and avoid the ```.unwrap()```, right?

xunilrj commented:
> I'm not sure what the rustiest way of implementing `parse_list` ...

What I see other parser combinators libs doing is:

https://docs.rs/nom/latest/nom/multi/fn.separated_list1.html
```rust
pub fn separated_list1<I, O, O2, E, F, G>(
    sep: G, 
    f: F
) -> impl FnMut(I) -> IResult<I, Vec<O>, E> 
where
    I: Clone + InputLength,
    F: Parser<I, O, E>,
    G: Parser<I, O2, E>,
    E: ParseError<I>
{
...
}
```

lovely if you hate one-letter-variables.... :stuck_out_tongue:


https://docs.rs/chumsky/0.6.0/chumsky/trait.Parser.html#method.separated_by

```rust
fn separated_by<U, P: Parser<I, U>>(self, other: P) -> SeparatedBy<Self, P, U>
where
    Self: Sized, 
{
    ...
}
```

I have the impression that ```nom``` style is closer to ours.

ematipico commented:
I implemented the proposed solution using `trait`

@MichaReiser : your proposed solution as `ParseList<T>` at the moment is not implemented. I need to review the architecture of the trait first.

If we decided to support `ParseList<ParsedSyntax>` and `ParseList<ConditionalSyntax>`, we'll need to implement an internal `type` and some functions (like `finish_list`) can't have a default implementation.

ematipico commented:
> I implemented the proposed solution using `trait`
> 
> @MichaReiser : your proposed solution as `ParseList<T>` at the moment is not implemented. I need to review the architecture of the trait first.
> 
> If we decided to support `ParseList<ParsedSyntax>` and `ParseList<ConditionalSyntax>`, we'll need to implement an internal `type` and some functions (like `finish_list`) can't have a default implementation.

Just implemented it using the `Self::FinishedSyntax` internal type.

Now the trait must implement `finish_list` method. Do you think we should implement it for `parse_element` too?

MichaReiser commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,103 @@
+/// An utility that gives finer control on how to parse a list of elements
+pub trait ParseList {
+	/// The type contained inside [ParsedSyntax]
+	type FinishedSyntax;
```
Oh, that's neat. I was mainly thinking to support `ParseList<T>` where `T` is the return type of `parse_element`.

MichaReiser commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,103 @@
+
+	/// When calling [parse_separated_list], this method checks, inside the loop, if the parser
+	/// is at a position where the current token is element that will separate the list
+	fn is_at_separating_element(&mut self, _p: &mut Parser) -> bool {
+		unimplemented!("When calling `parse_separated_list`, you need to implement this method.");
+	}
```
I would probably create a separate trait for parsing separated lists and then not implement this method.

ematipico commented on crates/rslint_parser/src/parser/parse_list.rs:
```diff
@@ -0,0 +1,103 @@
+
+	/// This method is used to check the current token inside the loop. When this method return [false],
+	/// the trait will exit from the loop.
+	fn is_at(&mut self, p: &mut Parser) -> bool;
```
The first time I read this proposal I thought it was syntax sugar to check if we are at the end of the file, which is not what you meant. Although I don't have other naming alternatives 😢

MichaReiser replied:
`is_at_list_end`?

ematipico replied:
Sounds good

MichaReiser commented on crates/rslint_parser/src/lib.rs:
```diff
@@ -337,13 +339,13 @@ pub trait SyntaxFeature: Sized {
 	/// Returns [Present(Invalid)], calls the `error_builder` to create a diagnostic,
 	/// adds the diagnostic to the parsing diagnostics, and returns [Present(Invalid)]
-	/// if the parsing context does support this syntax feature and the `syntax` is [Present].
+	/// if the parsing context does support this syntax feature and the `syntax` is [ParsedSyntax::Present].
```
Is this the proper syntax to use? Sad, that rust isn't telling me about syntax errors in docs.

ematipico replied:
You need to run `cargo doc` if you want to see the warnings. Yes, its' the correct syntax.

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,152 @@
+
+/// A generic that defines a generic behaviour for all the possible lists.
+///
+pub trait List {
```
Nit: Rust traits are often verbs, which is why I would call this `ParseList`. Or from another perspective, the trait itself isn't a list, it only knows how to parse a list.

ematipico replied:
`PreparedList` or `PrepareList`?

MichaReiser replied:
I personally would go with `ParseList`, also matches the main method which is called `parse_list`.

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,152 @@
+	/// Parses a single element of the list
+	fn parse_element(&mut self, p: &mut Parser) -> ParsedSyntax<Self::ParsedElement>;
+
+	/// The [SyntaxKind] used to name the list
+	fn list_kind(&mut self) -> SyntaxKind {
+		SyntaxKind::LIST
+	}
```
We may need to differentiate here. All our lists always contain a `LIST` node. For example, array expression contains a list with all the elements. I'm not sure if it ever makes sense to override this except if the `ParseList` also parses the start/end tokens of e.g. an `ArrayExpression` in which case the "list" would be the whole array expression node.

ematipico replied:
> I'm not sure if it ever makes sense to override

Thinking about cross parser/language, I thought that it would make sense to give the possibility to change the syntax kind. 

Are you saying that we would return the `Marker` instead of the `CompletedMarker` and let the consumer to wrap it into a `LIST` based on the situation?

MichaReiser replied:
LIST is probably going to be a language-agnostic kind, the same as EOF

A list can never be absent (it's either empty or has elements). This is why I'm not sure if it even makes sense to return anything from the `parse_list` method.

ematipico replied:
> A list can never be absent (it's either empty or has elements). This is why I'm not sure if it even makes sense to return anything from the parse_list method.

Oh that. I returned something for testing. I can revert the change.

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,152 @@
+		parsed_element: ParsedSyntax<Self::ParsedElement>,
+	) -> RecoveryResult;
+
+	/// [Diagnostic] thrown in case the parser is not able to recover
+	fn expected_element_error(p: &Parser, range: Range<usize>) -> Diagnostic;
```
Is this still needed since we don't have a default `recovery` implementation?

ematipico replied:
This is where it gets tricky. If not implemented, the `recover` method would panic. Same if `recovery()` is not implemented. How do we enforce this using the compiler? Is the documentation enough?

MichaReiser replied:
Not sure what you mean. I would remove this method as it's never needed by any implemented method of `ParseList`. Define the method in a specialised trait if any of them need it

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,152 @@
+	/// # Panics
+	///
+	/// It panics if the parser doesn't advance at each cycle of the loop
+	fn parse_list(&mut self, p: &mut Parser) -> ParsedSyntax<Self::ParsedList> {
+		let elements = self.start_list(p);
+		let mut progress = ParserProgress::default();
+		while !p.at(SyntaxKind::EOF) && !self.is_at_list_end(p) {
+			progress.assert_progressing(p);
+
+			let parsed_element = self.parse_element(p);
+
+			if self.recover(p, parsed_element).is_err() {
+				break;
+			}
+		}
+		self.finish_list(p, elements)
+	}
```
Should we unify the "normal" list with the "list" trait. Specialisations can still override `parse_list` to e.g. account for separators etc.

ematipico replied:
I'd like to keep it separate for now. There will be cases where we will have different separated list, I think.

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,152 @@
+	}
+
+	/// It creates a [ParsedSyntax] that will contain the list
+	fn finish_list(&mut self, p: &mut Parser, m: Marker) -> ParsedSyntax<Self::ParsedList>;
```
What's the reason that this isn't implemented, considering that the trait knows the list kind? 

Are there use cases where we need to override `start/finish_list`

It would be nice if `finish_list` creates a `missing` marker for empty lists. The facade should be able to handle empty lists.

ematipico replied:
> What's the reason that this isn't implemented, considering that the trait knows the list kind?

We don't know the type of `Self::ParsedList`, this means that we can't have a default implementation. 

We can't do

```rust
Present(m.complete(p, self.list_kind()))
```

Because the types don't match

> Are there use cases where we need to override start/finish_list

Probably in cases where the list is a `ConditionalSyntax`

> It would be nice if finish_list creates a missing marker for empty lists

We could enforce it but at the moment lists are not empty in the current implementations. When they are empty, they are still there.

MichaReiser replied:
OK, so how would this work. Does `recover` set an internal boolean flag to `invalid` if a syntax turned out to be invalid and then returns `Invalid`/`Valid` depending on that flag inside of `finish_list`? 

I'm still not sure if we should have `start/finish` list because all lists ever only are of the `kind` `LIST` and it makes it difficult to implement the optimisation to insert a `missing` slot for lists that contain no elements. 

There's further the point that `parse_list` returns `()`, so the caller has no access to the return value of `finish_list`. One could then argue that `parse_list` should return a `ParsedSyntax` but that enforces all callers to handle the `ParsedSyntax` which isn't necessarily for lists (as long as they are not `ConditionalSyntax`. 

I would go with the following:
* Change the return type of `parse_list` to `Option<TList>`
* Replace the `start_list`/`finish_list` with a `finish_list(Option<CompletedMarker>) -> Option<T>` that can decide if it wants to wrap the `CompletedMarker`. 

Now this approach has the downside that the compiler won't create a warning if `finish_list` returns `Option<ConditionalSyntax>`. 

How many use cases do we have where the parser returns a `ConditionalSyntax of List`? Is it worth adding all the complexity for this edge case?

ematipico replied:
I would suggest to keep it simple for now and make changes as we see fit - if we see fit - based on the usage in the future.

At the moment we don't have lists that have `ConditionalSyntax` (I haven't seen them). In case we will have the case, we will change the code base accordingly.

So we can return nothing - `()` - from `parse_list`. We remove `type ParsedList` and we always return a `Present(CompletedMarker)` inside `finish_list`, which will have a default implementation now.

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,152 @@
+	/// Tells the parser to mark the current token as missing, continuing the loop.
+	/// This function is used for [Self::parse_list].
+	fn parse_separating_element(&mut self, p: &mut Parser) {
+		p.missing();
```
This shouldn't add a missing marker but instead bump the separator token. 

We can probably unify the `parse_separator` and `is_at_separator` methods by introducing `expect_separator` which expects the separator token (and adds the missing marker)

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,152 @@
+
+			if self.is_at_separating_element(p) {
+				self.parse_separating_element(p);
+				continue;
```
You should continue here. Each separator should be followed by  an element except if we're at the end of the list, in which case you want to break
```suggestion
				if self.is_at_list_end(p) { 
				  break;
				}
```

MichaReiser commented on crates/rslint_parser/src/syntax/stmt.rs:
```diff
@@ -1054,6 +1054,47 @@ pub fn parse_for_statement(p: &mut Parser) -> ParsedSyntax<CompletedMarker> {
+	}
+
+	fn is_at_list_end(&mut self, p: &mut Parser) -> bool {
+		p.at_ts(token_set![T![default], T![case], T!['}']])
```
`default` / `case` don't mark the end of the list?

ematipico replied:
Sorry I didn't understand the comment

MichaReiser replied:
`is_at_list_end` tests if the list is at the end... But a `default` token doesn't signal the end of the list, it signals that the next element in the list is a default case.

MichaReiser commented on crates/rslint_parser/src/syntax/stmt.rs:
```diff
@@ -1054,6 +1054,47 @@ pub fn parse_for_statement(p: &mut Parser) -> ParsedSyntax<CompletedMarker> {
+	}
+
+	fn recovery() -> ParseRecovery {
+		ParseRecovery::new(JS_UNKNOWN_STATEMENT, STMT_RECOVERY_SET)
```
An `unknown` statement is no valid child for a switch clause.

MichaReiser commented on crates/rslint_parser/src/syntax/stmt.rs:
```diff
@@ -1054,6 +1054,43 @@ pub fn parse_for_statement(p: &mut Parser) -> ParsedSyntax<CompletedMarker> {
+	) -> parser::RecoveryResult {
+		parsed_element.or_recover(
+			p,
+			&ParseRecovery::new(JS_UNKNOWN_EXPRESSION, STMT_RECOVERY_SET),
```
`JS_UNKOWN_STATEMENT`... and I finally understood that this is parsing the statements, not the switch cases... sorry

ematipico replied:
no problem!

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,156 @@
+	/// The type of syntax that will be returned by [Self::parse_list]. The type will be the generic for [ParsedSyntax]
+	type ParsedList;
+
+	/// Tells the parser to parse the current token, continuing the loop.
+	/// This function is used in [Self::parse_list].
+	fn parse_separating_element(&mut self, p: &mut Parser) {
+		// bump the separator
+		p.bump_any();
+	}
+
+	/// The [SyntaxKind] of the element that separates the elements of the list
+	fn separating_element_kind(&mut self) -> SyntaxKind;
+
+	/// Method called at each iteration of the the loop and checks if the expected
+	/// separator is present.
+	///
+	/// If present, it [parses](Self::parse_separating_element) it and continues with loop.
+	/// If not present, it adds a missing marker.
+	fn expect_separator(&mut self, p: &mut Parser) -> bool {
+		if p.expect_required(self.separating_element_kind()) {
+			self.parse_separating_element(p);
+			true
+		} else {
+			false
+		}
+	}
```
Do we need all three +  `is_at_separating_element`? `p.expect_required` eats the separator element. and also returns `true`/`false`. I think having just `expect_separator` should be enough that implementations of this trait must provide (they may decide to have some more complicated logic, for example, to re-map the kind.

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,156 @@
+				break;
+			}
+		}
+		self.finish_list(p, elements).unwrap();
```
`finish_list` shouldn't return `ParsedSyntax` if you call `unwrap` on it anyway. Make it so that it returns `CompletedMarker` or `InvalidSyntax` instead

MichaReiser commented on crates/rslint_parser/src/parser/parse_lists.rs:
```diff
@@ -0,0 +1,156 @@
+
+/// A generic that defines a generic behaviour for all the possible lists.
+///
+pub trait ParseList {
+	/// The type returned when calling the function [Self::parse_element]
+	type ParsedElement;
+
+	/// Parses a single element of the list
+	fn parse_element(&mut self, p: &mut Parser) -> ParsedSyntax<Self::ParsedElement>;
+
+	/// It creates a marker just before starting a list
+	fn start_list(&mut self, p: &mut Parser) -> Marker {
+		p.start()
+	}
+
+	/// This method is used to check the current token inside the loop. When this method return [false],
+	/// the trait will exit from the loop.
+	///
+	/// Usually here you want to check the current token.
+	fn is_at_list_end(&mut self, p: &mut Parser) -> bool;
+
+	/// This method is used to recover the parser in case [Self::parse_element] returns [ParsedSyntax::Absent]
+	fn recover(
+		&mut self,
+		p: &mut Parser,
+		parsed_element: ParsedSyntax<Self::ParsedElement>,
+	) -> RecoveryResult;
+}
```
Should we remove this trait? It doesn't provide any useful methods on its own.

ematipico replied:
It does, it provides shared behaviour among traits

MichaReiser reviewed (commented):
I realised that I had a misunderstanding of how `traits` work in rust and am now wondering if `ParseNormaleList` and `ParseSeparatedList` should be structs instead... 

```rust
trait ParseList {
  type TList;
  type TItem;

  fn parse_list(&mut self, p: &mut Parser) -> Option<self::TList> {
    // default implementation
  }

  fn parse_item(&mut self, p: &mut Parser) -> self::TItem;

  fn recover(&mut self, p: &mut Parser, item: self::TItem) -> RecoveryResult<self::TItem>;
}

struct NormalList<TItem, Parse> {
  parse_fn: Parse
};

impl<Parse> NormalList<CompletedMarker, P> where P: Fn(&mut Parser) -> ParsedSyntax<CompletedMarker> { ... }

impl<Parse> NormalList<ConditionalSyntax, P> where P: Fn(&mut Parser) -> ParsedSyntax<ConditionalSyntax> { 
}

fn parse_list<P>(p: &mut Parser, parse_item: P, recovery: ParseRecovery) -> Option<CompletedMarker> where P: Fn(&mut Parser) -> Option<CompletedMarker> {
   NormalList { parse: parse_item }.parse_list(p)
 }
}

#[must_use])
fn parse_conditional_list<P>(p: &mut Parser, parse_item: P, recovery: ParseRecovery) -> Option<ConditionalSyntax> where P: Fn(&mut Parser) -> Option<ConditionalSyntax> {
   NormalList { parse: parse_item }.parse_list(p)
 }
}
```

The structs would then become implementation details of the `parse_list` functions. 

I'm not sure of what the downside of this are compared to use multiple traits.

MichaReiser reviewed (approved):
Amazing! It will allow me to delete so much code from the module implementation :)
